main.rb

```diff
@@ -1,44 +1,261 @@
+        yield to_a[a]
         a += 1
       end
-      self 
-    else 
+      self
+    else
       to_enum(:my_each)
-    end  
+    end
   end
-  
-# 2.my_each_index 
+
+  # 2.my_each_index
   def my_each_with_index
-    #return to_enum(:my_each_index) unless block_given?
+    # return to_enum(:my_each_index) unless block_given?
     if block_given?
       a = 0
-      while a < self.length
-        yield(self[a], a)
+      while a < to_a.length
+        yield(to_a[a], a)
         a += 1
       end
-      self 
-    else 
+      self
+    else
       to_enum(:my_each_with_index)
-    end  
+    end
   end
 
-# 2.my_select
+  # 2.my_select
   def my_select
-    #return to_enum(:my_each_index) unless block_given?
+    # return to_enum(:my_each_index) unless block_given?
     if block_given?
-      newArr = []
-      self.my_each {|item| newArr << item if yield(item)}
-      newArr
-    end 
-  end  
+      new_arr = []
+      to_a.my_each { |item| new_arr << item if yield(item) }
+      new_arr
+    else
+      to_enum(:my_each_with_index)
+    end
+  end
+
+  # 4.my_all?
+  def my_all?(argument = nil)
+    index = 0
+    my = to_a
+    while index < my.length
+      if block_given?
+        return false if yield(my[index]) == false
+      else
+        if argument.nil? # when there is no argument
+          return false if my[index].nil? || my[index] == false
+        elsif argument.class == Regexp # when there is an argument
+          return false if my[index].match?(argument) != true
+        elsif argument.class == Class
+          return false if my[index].is_a?(argument) != true
+        else # when the argument is no regexp ni class
+          return false if my[index] != argument
+        end
+      end
+      index += 1
+    end
+    true
+  end
+
+  # 4.my_any?
+  def my_any?(arg = nil)
+    a = 0
+    o = false
+    my = to_a
+    if block_given?
+      while a < my.length
+        if yield(my[a]) == true
+          o = true
+          break
+        end
+        a += 1
+      end
+      o
+    elsif arg.nil?
+      while a < my.length
+        p 'a'
+        if my[a] != false
+          o = true
+          break
+        elsif my[a].nil?
+          o = true
+        end
+        a += 1
+      end
+      o
+    elsif !arg.nil? && arg.class == Class
+      o = false
+      while a < my.length
+        if (my[a].is_a? arg) == true
+          o = true
+          break
+        end
+        a += 1
+      end
+      o
+    elsif arg.class == Regexp
+      while a < my.length
+        if !my[a].match(arg).nil?
+          o = true
+          break
+        else
+          o = false
+        end
+        a += 1
+      end
+      o
+    elsif arg.class != Regexp && arg.class == String
+      while a < my.length
+        if my[a] == arg
+          o = true
+          break
+        else
+          o = false
+        end
+        a += 1
+      end
+      o
+    elsif arg.class == Array
+      p 'aaa'
+    else
+      true
+    end
+  end
```

Comment:
- [ ] **OPTIONAL** You can make your my_any? way shorter by using this: (also applies to my_all? and my_none?)

```
def my_any? (arg = nil)
  if block_given?
    # your code
  elsif arg.nil?
    # my_any? { # add a block } <----------- this way you can enter to the logic you already did in if block_given?
  else
    # my_any? { # add a block using the argument and investigate the === operator }
  end
end
```